### DISCUSSION TITLE
Missing command class parameter for zwave keypad iSurpass J1825

### DISCUSSION THREAD

**Discussion by yoguiti** [zwave-js/zwave-js#6797] [Request Support / Investigate Issue]
### Checklist

- [X] I have read and followed the above instructions
- [X] I have checked the [troubleshooting section](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index) and my problem is **not** described there.

- [X] I have read the [changelog](https://github.com/zwave-js/node-zwave-js/blob/master/CHANGELOG.md) and my problem was **not** mentioned there or the fix **did not work**.

### Describe the issue

What is happening?
After inclusion of device j1825 (keypad isurpass), the following command class parameter is not displayed : 
Command Class : 113 (0x71)
Command Index : 153
Type : Byte (int)
Name : Memory Code User

Missing in the below picture :
![image](https://github.com/zwave-js/node-zwave-js/assets/10196037/5d184ec8-6cde-46ff-b6f2-08069289cff5)

This happens on ZwaveJS (I am using Jeedom). On other Zwave implementation (previous jeedom zwave plugin), this was correctly recognized (see picture attached)
![image](https://github.com/zwave-js/node-zwave-js/assets/10196037/2c9687e8-16ba-445f-954f-653649fceab3)

Anything else we should know?
I do not know how to modify the j1825.json to correct this. Any help would be greatly appreciated. The device json is here :

![image](https://github.com/zwave-js/node-zwave-js/assets/10196037/30dd9f79-01d0-4868-89c3-683b0ed44bf6)

### Software versions

Z-Wave JS UI: 9.6.2

### Device information

Manufacturer: Shenzhen iSurpass Technology Co., Ltd
Model name: J1825 Wall Reader
Node ID: 95
Firmware : 2.16

### Checklist

- [X] I [made sure](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index?id=rule-1-use-the-driver-log-luke) to provide a **driver log** on level **debug**.

- [ ] The log includes a re-interview of the problematic device _(if applicable)_.

- [ ] The log includes the problematic interaction with the device _(if applicable)_.

- [ ] I provided the node ID of the problematic device _(if applicable)_.

### Upload Logfile
The log is here, with device ID 095
[zwavejs_2024-04-26.log](https://github.com/zwave-js/node-zwave-js/files/15131735/zwavejs_2024-04-26.log)

I have actionned the device, and the log code which is relevant is below. I am looking to get the "UserId" parameter :

"2024-04-26T14:02:34.462Z CNTRLR   [Node 095] [Notification] alarmType: metadata updated             [Endpoint 0]
2024-04-26T14:02:34.463Z CNTRLR   [Node 095] [+] [Notification] alarmType: 0                        [Endpoint 0]
2024-04-26T14:02:34.463Z CNTRLR   [Node 095] [Notification] alarmLevel: metadata updated            [Endpoint 0]
2024-04-26T14:02:34.463Z CNTRLR   [Node 095] [+] [Notification] alarmLevel: 0                       [Endpoint 0]
2024-04-26T14:02:34.464Z DRIVER « [Node 095] [REQ] [ApplicationCommand]
                                  └─[SecurityCCCommandEncapsulation]
                                    │ sequenced: false
                                    └─[NotificationCCReport]
                                        notification type:   Access Control
                                        notification status: 255
                                        notification event:  Keypad unlock operation
                                        event parameters:
                                          userId: 0x01
2024-04-26T14:02:34.464Z CNTRLR   [Node 095] [~] [Door Lock] currentMode: 255 => 0                  [Endpoint 0]
2024-04-26T14:02:34.465Z CNTRLR   [Node 095] [Notification]
                                    type:   Access Control
                                    event:  Keypad unlock operation
                                    userId: 1

[a comment by an automated account is not shown]

**Reply by yoguiti** [reply]
driver log is irrelevant to the issue.

**Comment by AlCalzone** [comment]
I'm not sure what that value is supposed to be. If it's a notification, it's not a one from the Z-Wave specifications and apparently not one the device even reports to be supported.

So yes, the log is relevant and it should include some event that would modify this value.

**Reply by yoguiti** [reply]
Thanks a lot, I have added the log file above, and I think the relevant code is here : 

2024-04-26T14:02:34.462Z CNTRLR   [Node 095] [Notification] alarmType: metadata updated             [Endpoint 0]
2024-04-26T14:02:34.463Z CNTRLR   [Node 095] [+] [Notification] alarmType: 0                        [Endpoint 0]
2024-04-26T14:02:34.463Z CNTRLR   [Node 095] [Notification] alarmLevel: metadata updated            [Endpoint 0]
2024-04-26T14:02:34.463Z CNTRLR   [Node 095] [+] [Notification] alarmLevel: 0                       [Endpoint 0]
2024-04-26T14:02:34.464Z DRIVER « [Node 095] [REQ] [ApplicationCommand]
                                  └─[SecurityCCCommandEncapsulation]
                                    │ sequenced: false
                                    └─[NotificationCCReport]
                                        notification type:   Access Control
                                        notification status: 255
                                        notification event:  Keypad unlock operation
                                        event parameters:
                                          userId: 0x01
2024-04-26T14:02:34.464Z CNTRLR   [Node 095] [~] [Door Lock] currentMode: 255 => 0                  [Endpoint 0]
2024-04-26T14:02:34.465Z CNTRLR   [Node 095] [Notification]
                                    type:   Access Control
                                    event:  Keypad unlock operation
                                    userId: 1


I am looking for the userId =1 that is sent with the notification.

Thanks a lot for your help !

**Reply by AlCalzone** [reply]
I'm not sure how Domoticz exposes that information. AFAIK it uses the MQTT API. @robertsLando do you know how the userId event parameter (or notification event parameters in general) are exposed in MQTT?

**Reply by robertsLando** [reply]
This is the code I use to handle node notifications:

https://github.com/zwave-js/zwave-js-ui/blob/c37ae2ad368013715bf894403703208d170cc756/api/lib/ZwaveClient.ts#L5596-L5627

Seems I don't have a specific `if` case for `Access Control`, is that the payload expected in case of a notification? I mean:

```
type: Access Control
event: Keypad unlock operation
userId: 1
```

**Reply by yoguiti** [reply]
Thanks a lot @robertsLando to help me solve this. I have no idea unfortunately, I put the log I had from zwave js here : 
https://github.com/zwave-js/node-zwave-js/files/15131735/zwavejs_2024-04-26.log

Do not hesitate to tell if you want me to do something to test. I am using jeedom with zwaveJS.

I was hopping there would be a way to modify the file 1825.js below to catch this parameter.
https://github.com/zwave-js/node-zwave-js/blob/bf8cfa8a4ae5f3365c3b9c5fbb679350f2104ed9/packages/config/config/devices/0x021c/j1825.json#L4

In fact I used to use OpenZwave before and it was working. I have just looked into the source code of openzwave and found the file that was used to get this value I think (I added ".txt at the end of the file) :
[Alarm.cpp.txt](https://github.com/zwave-js/node-zwave-js/files/15185186/Alarm.cpp.txt)

The relevant part is here, and they create a command named "Memory Code User" => This is the UserId you are looking for, they just name it differently in openzwave : 

         if( AlarmCmd_SupportedReport == (AlarmCmd)_data[0] )                                                                                                                                                                                      
        {                                                                                                                                                                                                                                         
                if( Node* node = GetNodeUnsafe() )                                                                                                                                                                                                
                {                                                                                                                                                                                                                                 
                        // We have received the supported alarm types from the Z-Wave device                                                                                                                                                      
                        Log::Write( LogLevel_Info, GetNodeId(), "Received supported alarm types" );                                                                                                                                               
                                                                                                                                                                                                                                                  
                        node->CreateValueByte( ValueID::ValueGenre_User, GetCommandClassId(), _instance, AlarmIndex_SourceNodeId, "SourceNodeId", "", true, false, 0, 0 );                                                                        
                        Log::Write( LogLevel_Info, GetNodeId(), "    Added alarm SourceNodeId" );                                                                                                                                                 
                                                                                                                                                                                                                                                  
                        // Parse the data for the supported alarm types                                                                                                                                                                           
                        uint8 numBytes = _data[1];                                                                                                                                                                                                
                        for( uint32 i=0; i<numBytes; ++i )                                                                                                                                                                                        
                        {                                                                                                                                                                                                                         
                                for( int32 bit=0; bit<8; ++bit )                                                                                                                                                                                  
                                {                                                                                                                                                                                                                 
                                        if( ( _data[i+2] & (1<<bit) ) != 0 )                                                                                                                                                                      
                                        {                                                                                                                                                                                                         
                                                int32 index = (int32)(i<<3) + bit;                                                                                                                                                                
                                                if( index < Alarm_Count )                                                                                                                                                                         
                                                {                                                                                                                                                                                                 
                                                        node->CreateValueByte( ValueID::ValueGenre_User, GetCommandClassId(), _instance, index+3, c_alarmTypeName[index], "", true, false, 0, 0 );                                                
                                                        Log::Write( LogLevel_Info, GetNodeId(), "    Added alarm type: %s", c_alarmTypeName[index] );                                                                                             
                                                        if (index == 6)                                                                                                                                                                           
                                                        {                                                                                                                                                                                         
                                                                node->CreateValueByte( ValueID::ValueGenre_User, GetCommandClassId(), _instance, 0x99, "Memory Code User", "", true, false, 0, 0 );                                      
                                                        }
                                                } else {                                                                                                                                                                                          
                                                        Log::Write( LogLevel_Info, GetNodeId(), "    Unknown alarm type: %d", index );                                                                                                            
                                                }                                                                                                                                                                                                 
                                        }                                                                                                                                                                                                         
                                }                                                                                                                                                                                                                 
                        }                                                                                                                                                                                                                         
                }                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                  
                ClearStaticRequest( StaticRequest_Values );                                                                                                                                                                                       
                return true;                                                                                                                                                                                                                      
        }

**Reply by robertsLando** [reply]
@yoguiti The question was for @AlCalzone :) Maybe I have to do some changes on my side too

**Reply by AlCalzone** [reply]
> Seems I don't have a specific if case for Access Control

There shouldn't be a special case just for Access Control. Notifications may or may not have event parameters attached. If they have, the parameters should be exposed.

**Reply by robertsLando** [reply]
I actually use this to parse notification parameters: 

https://github.com/zwave-js/zwave-js-ui/blob/e51cf52bea1d1795f75fdf68abfbe9568131cfaf/api/lib/ZwaveClient.ts#L6413-L6420 

I think that should return an object with the userId value like `{userId: 0x01}` ? If so that is the payload of the mqtt message sent to the node topic / notificationCC

**Reply by AlCalzone** [reply]
Ok then it might be up to Domotics to actually parse and expose that info.

**Reply by yoguiti** [reply]
Thanks a lot for your support.
However I am sorry I am unsure what needs to be done. Could you please clarify ? This userId value does not appear in zwave-js UI. Is there a modification to be made so it is the case ? Anything I can test ?

Thanks

**Reply by yoguiti** [reply]
@AlCalzone , when you say it might be up to domotics to parse and expose this info, you mean it is sent by ZwaveJS but the receiver end (Jeedom / Domotics) don't parse it correctly ? If that is what you are saying I am not sure this is the case because it does not appear either in zwaveJS UI. I would realla appreciate your help to move forward, and happy to test something. Otherwise I will have to do stupidely ugly like parsing the log of ZwaveJS in live to catch this event, which I definitely don't want to do ....

Thanks again !

**Reply by AlCalzone** [reply]
Z-Wave JS doesn't create values for event metadata that only has a meaning the instant it is received.
You should see the information in the events tab, which means Domotics/Jeedom should also see the information.
